plugins/bulk_upload/drop_folder_xml/plugins/providers/ap_feed/batch/KApFeedDropFolderEngine.php

```diff
@@ -0,0 +1,150 @@
+<?php
+/**
+ * @package plugins.ApFeedDropFolder
+ */
+class KApFeedDropFolderEngine extends KFeedDropFolderEngine
+{
+	/**
+	 * The API Key for the AP feed being processed.
+	 * @var string
+	 */
+	protected $apiKey;
+	
+	const AP_FEED_PAGE_SIZE = 100;
+	
+	public function watchFolder(KalturaDropFolder $dropFolder)
+	{
+		/* @var $dropFolder KalturaApFeedDropFolder */
+		KalturaLog::info("Watching drop folder with ID [" . $dropFolder->id . "]");
+		$this->dropFolder = $dropFolder;
+		$this->apiKey = $dropFolder->apApiKey;
+		
+		$feedUrl = $this->dropFolder->path;
+		
+		$existingDropFolderFilesMap = $this->loadDropFolderFiles();
+		
+		$counter = 0;
+		$break = false;
+		
+		do{
+			
+			//Get Drop Folder feed and import it into an array
+			$feedContent = $this->fetchFeedContent($feedUrl);
+			$feed = json_decode($feedContent, true);
+			if (!$feed)
+			{
+				KalturaLog::info('Feed page could not be displayed, no way to continue. Breaking out of the process.');
+				break;
+			}
+			$lastPageCount = $feed['data']['current_item_count'];
+			
+			$feedItems = $feed['data']['items'];
+			
+			foreach ($feedItems as $part)
+			{
+				
+				if ($counter > intval ( KBatchBase::$taskConfig->params->mrss->limitProcessEachRun))
+				{
+					KalturaLog::info('Process limit reached.');
+					$break = true;
+					break;
+				}
+				
+				$datum = $part['item'];
+				$feedItemJson = json_decode($this->fetchFeedContent($datum['uri']), true);
+				if (!$feedItemJson)
+				{
+					KalturaLog::info('Feed item could not be retrieved, continue to the next one.');
+					continue;
+				}
+				
+				$feedItem = new DOMDocument();
+				$feedItem->loadXML('<item/>');
+				
+				$this->arrayToXml($feedItemJson['data']['item'], $feedItem->documentElement);
+				
+				KalturaLog::info ('Single item: ' . print_r($feedItem->saveXML(), true));
+				
+				$counter += $this->watchProcessSingleItem(simplexml_import_dom($feedItem), $existingDropFolderFilesMap);
+				
+			}
+			
+			if($break)
+			{
+				KalturaLog::info('Finished current run.');
+				break;
+			}
+			
+			if (isset($feed['data']['next_page']))
+			{
+				$feedUrl = $feed['data']['next_page'];
+			}
+			
+		}while ($lastPageCount == self::AP_FEED_PAGE_SIZE);
```

Comment:
This is a weak end case, since it is possible that due to changes that happen in parallel, one of the pages will return less than page size, still there will be more to come.
Safer approach will be until lastPageCount == 0.

Reply:
@MosheMaorKaltura  because this is a professionally generated news feed, odds are slim that we could catch it just at the time of day it gets updated, which is exactly why I cut off the interaction with it on what is perceived as the last page. Also I don't see how the same case can't happen even if the last page item count **is** 0. This interaction isn't a one-off - this is a drop folder pull, so it gets repeated periodically. Any content that was not ingested on a single run will likely get pulled in on the next one.